sw/airborne/modules/calibration/mag_calib_ukf.c

```diff
@@ -0,0 +1,206 @@
+    TRICAL_norm_set(&mag_calib, MAG_CALIB_UKF_NORM);
+    TRICAL_noise_set(&mag_calib, MAG_CALIB_UKF_NOISE_RMS);
+    VERBOSE_PRINT("magnetometer initial calibration(x_n: %d  y_n: %d  z_n: %d)\n", _imu->mag_neutral.x, _imu->mag_neutral.y, _imu->mag_neutral.z);
+    _imu->mag_neutral.x = 0;
+    _imu->mag_neutral.y = 0;
+    _imu->mag_neutral.z = 0;
+    /* Delft */
+    H.x = AHRS_H_X;
+    H.y = AHRS_H_Y;
+    H.z = AHRS_H_Z;
+    VERBOSE_PRINT("Local magnetic field loaded from airframe file (Hx: %4.2f, Hy: %4.2f, Hz: %4.2f)\n", H.x, H.y, H.z);
+
+#if MAG_CALIB_UKF_HOTSTART
+    snprintf(hotstart_file_name, 512, "%s", STRINGIFY(MAG_CALIB_UKF_HOTSTART_SAVE_FILE));
+    mag_calib_hotstart_read();
+#endif
+#endif
+}
+
+void mag_calib_ukf_run(struct Imu *_imu) {
+#if USE_MAGNETOMETER
+    float meas[3] = {0.0, 0.0, 0.0}, calib_meas[3] = {0.0, 0.0, 0.0};
+    /** Update geo_mag based on MAG_CALIB_UKF_GEO_MAG_TIMEOUT (0 = no periodic updates) **/
+    if(update_geo_mag_field && geo_mag.ready){
+        double n                = double_vect3_norm(&geo_mag.vect);
+        H.x                     = (float) geo_mag.vect.x / n;
+        H.y                     = (float) geo_mag.vect.y / n;
+        H.z                     = (float) geo_mag.vect.z / n;
+        mag_field_from_geo_mag  = time(0);
+        update_geo_mag_field    = false;
+        VERBOSE_PRINT("Updating local magnetic field from geo_mag module (Hx: %4.2f, Hy: %4.2f, Hz: %4.2f)\n", H.x, H.y, H.z);
+    }
+    if(MAG_CALIB_UKF_GEO_MAG_TIMEOUT && GpsFixValid() && difftime( time(0), mag_field_from_geo_mag ) >= MAG_CALIB_UKF_GEO_MAG_TIMEOUT){
+        geo_mag.ready           = false;
+        geo_mag.calc_once       = true;   ///< Geo_mag will not re-update the calculation when the throttle is on so this is neccesary
+        update_geo_mag_field    = true;
+    }
+    /** Update magnetometer UKF and calibrate measurement **/
+    if(_imu->mag.x != 0 || _imu->mag.y != 0 || _imu->mag.z != 0){
+        /** Rotate the local magnetic field by our current attitude **/
+        struct FloatQuat* body_quat = stateGetNedToBodyQuat_f();
+        struct FloatRMat body_rmat;
+        float_rmat_of_quat(&body_rmat, body_quat);
+        struct FloatVect3 expected_meas;
+        float_rmat_vmult(&expected_meas, &body_rmat, &H);
+        float expected_mag_field[3] = { expected_meas.x, expected_meas.y, expected_meas.z };
+        /** Update magnetometer UKF **/
+        meas[0] = MAG_FLOAT_OF_BFP( _imu->mag.x );
+        meas[1] = MAG_FLOAT_OF_BFP( _imu->mag.y );
+        meas[2] = MAG_FLOAT_OF_BFP( _imu->mag.z );
+        TRICAL_estimate_update(&mag_calib, meas, expected_mag_field);
+        TRICAL_measurement_calibrate(&mag_calib, meas, calib_meas);
+        /** Save calibrated result **/
+        _imu->mag.x = (int32_t) MAG_BFP_OF_REAL( calib_meas[0] );
+        _imu->mag.y = (int32_t) MAG_BFP_OF_REAL( calib_meas[1] );
+        _imu->mag.z = (int32_t) MAG_BFP_OF_REAL( calib_meas[2] );
```

Comment:
Ah ok, so it probably works since subsystems are initialized before modules.

However if I send another abi message after processing the one from the IMU won't the AHRS see duplicate magnetometer messages if one of the following defines isn't correctly alterred:
`AHRS_FC_MAG_ID`, `AHRS_ICE_MAG_ID`, `AHRS_MLKF_MAG_ID`, `AHRS_FINV_MAG_ID`,  `AHRS_DCM_MAG_ID`, `AHRS_ICQ_MAG_ID`, `INS_FINV_MAG_ID`

I've tried to resolve this by hardcoding the defines for all current AHRS systems in  my header file:

```
#ifdef AHRS_FLOAT_CMPL_WRAPPER_H
#undef AHRS_FC_MAG_ID
#define AHRS_FC_MAG_ID MAG_CALIB_UKF_ID
#endif

#ifdef AHRS_INT_CMPL_EULER_WRAPPER_H
#undef  AHRS_ICE_MAG_ID
#define AHRS_ICE_MAG_ID MAG_CALIB_UKF_ID
#endif

#ifdef AHRS_FLOAT_MLKF_WRAPPER_H
#undef  AHRS_MLKF_MAG_ID
#define AHRS_MLKF_MAG_ID MAG_CALIB_UKF_ID
#endif

#ifdef AHRS_FLOAT_INVARIANT_WRAPPER_H
#undef  AHRS_FINV_MAG_ID
#define AHRS_FINV_MAG_ID MAG_CALIB_UKF_ID
#endif

#ifdef AHRS_FLOAT_DCM_WRAPPER_H
#undef  AHRS_DCM_MAG_ID
#define AHRS_DCM_MAG_ID MAG_CALIB_UKF_ID
#endif

#ifdef AHRS_INT_CMPL_QUAT_WRAPPER_H
#undef  AHRS_ICQ_MAG_ID
#define AHRS_ICQ_MAG_ID MAG_CALIB_UKF_ID
#endif

#ifdef INS_FLOAT_INVARIANT_WRAPPER_H
#undef  INS_FINV_MAG_ID
#define INS_FINV_MAG_ID MAG_CALIB_UKF_ID
#endif
```

is this acceptable?